realm/realm-jni/src/io_realm_internal_TableQuery.cpp

```diff
@@ -1146,12 +1150,20 @@ JNIEXPORT jlongArray JNICALL Java_io_realm_internal_TableQuery_nativeBatchUpdate
         std::vector<jlong> exported_handover_tableview_array(number_of_queries);
 
         // Step1: Position the shared group at the handover query version so we can import all queries
-
         // read the first query to determine the version we should use
         SharedGroup::Handover<Query> *handoverQueryPtr = HO(Query, handover_queries_pointer_array[0]);
         std::unique_ptr<SharedGroup::Handover<Query>> handoverQuery(handoverQueryPtr);
-        // position this shared group at the specified version
-        SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+
+        // if the SharedGroup is not in Read Transaction, we position it at the same version as the handover
+        // The Handover object doesn't prevent a SharedGroup version from no longer being accessible. In rare
+        // cases this means that the version in the Handover object is invalid and Realm Core will throw a
+        // BadVersion as result.
+        if (SG(bgSharedGroupPtr)->get_transact_stage() != SharedGroup::transact_Reading) {
+            SG(bgSharedGroupPtr)->begin_read(handoverQuery->version);
+        } else if (SG(bgSharedGroupPtr)->get_version_of_current_transaction() != handoverQuery->version) {
```

Comment:
@cmelchior I am not trying to review. @finnschiermer says 
> I don't think Core promises this call to be well defined, when used outside a transaction. It should be called from within a transaction, but see below for a bigger problem.

You can see his comment in my unit test for new `distinct()` submitted [here](https://github.com/realm/realm-core/commit/078e61244aa1551b7dcf475c4fa648e1dd62585c#diff-dfb86ea2c75fcee0e86ed82cdc5ae7adR11129)